polars/polars-core/src/frame/row/transpose.rs

```diff
@@ -54,25 +61,56 @@ impl DataFrame {
                 }
                 let cols = buffers
                     .into_iter()
-                    .enumerate()
-                    .map(|(i, buf)| {
-                        let mut s = buf.into_series().cast(dtype).unwrap();
-                        s.rename(&format!("column_{i}"));
-                        s
-                    })
+                    .map(|buf| buf.into_series().cast(dtype).unwrap())
                     .collect::<Vec<_>>();
                 Ok(DataFrame::new_no_checks(cols))
             }
+        }?;
+        out.set_column_names(colnames)?;
+        match keep_names_as {
+            Some(cn) => {
+                let namecol = Utf8Chunked::new(cn, self.get_column_names());
+                out.insert_at_idx_no_name_check(0, namecol.into()).cloned()
+            }
+            None => Ok(out),
         }
     }
 
     /// Transpose a DataFrame. This is a very expensive operation.
-    pub fn transpose(&self) -> PolarsResult<DataFrame> {
+    pub fn transpose(
+        &self,
+        keep_names_as: Option<&str>,
+        column_names: Option<Either<String, Vec<String>>>,
+    ) -> PolarsResult<DataFrame> {
+        let mut df = self.clone(); // Must be owned so we get the same type if dropping a column.
```

Comment:
You can put it behind a `Cow` then we don't need to clone.

Reply:
You heap allocate a new `Vec` and atomically increment every series and follow the indirections. DataFrames can be very wide.

Putting it behind a `Cow` is almost free in comparison.

Reply:
I guess I was working under the assumption no one would be transposing something with more than, say, a million columns, which would make this small in comparison to the rest of the operation. But it sounds easy enough to avoid moooooving around data we don't need to.

Reply:
Well the `Cow` is almost free, so it is good practice.

Reply:
Tell that to people from cultures where the cow is a symbol of wealth and prestige :-)